File: ReportGenerator.cpp
#include "ReportGenerator.hpp"
#include <iostream>
#include <algorithm>
#include <vector>
#include <string>

using namespace std;

// public methods
void ReportGenerator::generateActiveLoansReport(const std::vector<Transaction>& transactions,
    const vector<User>& users,
    const vector<Book>& books) const
{
    cout << "\n--- ОТЧЕТ ОБ АКТИВНЫХ ВЫДАЧАХ ---" << endl;
    bool found = false;
    for (const auto& t : transactions) {
        if (t.isActiveStatus()) {
            found = true;
            cout << "Транзакция ID: " << t.getTransactionId() << endl;

            // User search
            auto userIt = find_if(users.begin(), users.end(),
                [t](const User& u) { return u.getId() == t.getUserId(); });

            if (userIt != users.end()) {
                cout << "  Пользователь: " << userIt->getName() << " (ID: " << t.getUserId() << ")" << endl;
            }
            else {
                cout << "  Пользователь: ID " << t.getUserId() << " (Не найден)" << endl;
            }

            // Book search
            auto bookIt = find_if(books.begin(), books.end(),
                [t](const Book& b) { return b.getId() == t.getBookId(); });

            if (bookIt != books.end()) {
                cout << "  Книга: " << bookIt->getTitle() << " (ID: " << t.getBookId() << ")" << endl;
                cout << "  Срок возврата: " << t.getDueDate() << endl;
            }
            else {
                cout << "  Книга: ID " << t.getBookId() << " (Не найдена)" << endl;
            }
            cout << "-----------------------------------" << endl;
        }
    }
    if (!found) {
        cout << "Активных выданных книг нет." << endl;
    }
}

File: CourceWork2Java/src/ReportGenerator.java
import java.util.List;
import java.util.Optional;

public class ReportGenerator {

    // Constructors
    public ReportGenerator() {
    }

    // Public methods
    public void generateActiveLoansReport(List<Transaction> transactions, 
            List<User> users, List<Book> books) {
        System.out.println("\n--- ОТЧЕТ ОБ АКТИВНЫХ ВЫДАЧАХ ---");
        boolean found = false;
        for (Transaction t : transactions) {
        	if (t.isActiveStatus()) {
        		found = true;
        		System.out.println("Транзакция ID: " + t.getId());
        		
        		Optional<User> userOpt = users.stream().filter(u -> u.getId() == t.getUserId()).findFirst();

				userOpt.ifPresent(user -> System.out.println("  Пользователь: " + user.getName() + " (ID: " + t.getUserId() + ")"));
				
				Optional<Book> bookOpt = books.stream().filter(b -> b.getId() == t.getBookId()).findFirst();
				
				bookOpt.ifPresent(book -> {
				    System.out.println("  Книга: " + book.getTitle() + " (ID: " + t.getBookId() + ")");
				    System.out.println("  Срок возврата: " + t.getDueDate());
				});
				System.out.println("-----------------------------------");
        	}
        }
        if (!found) {
            System.out.println("Активных выданных книг нет.");
        }
    }
}